assets/js/base/hooks/customer/use-customer-data.js

```diff
@@ -19,73 +20,94 @@ export const useCustomerData = () => {
+	const [ debouncedCustomerData ] = useDebounce( customerData, 400, {
+		equalityFn: ( prevData, newData ) => {
+			return ! (
+				isShallowEqual( prevData.billingData, newData.billingData ) ||
+				isShallowEqual(
+					prevData.shippingAddress,
+					newData.shippingAddress
+				)
+			);
+		},
+	} );
+
+	/**
+	 * Set billing data.
+	 *
+	 * Contains special handling for email and phone so those fields are not overwritten if simply updating address.
+	 */
 	const setBillingData = useCallback( ( newData ) => {
-		setCustomerData( ( prevState ) => ( {
-			...prevState,
-			billingData: {
-				...prevState.billingData,
-				...newData,
-			},
-		} ) );
+		setCustomerData( ( prevState ) => {
+			return {
+				...prevState,
+				billingData: {
+					...prevState.billingData,
+					...newData,
+				},
+			};
+		} );
 	}, [] );
 
 	const setShippingAddress = useCallback( ( newData ) => {
 		setCustomerData( ( prevState ) => ( {
 			...prevState,
-			shippingAddress: {
-				...prevState.shippingAddress,
-				...newData,
-			},
+			shippingAddress: newData,
 		} ) );
 	}, [] );
 
-	// This will push both billing and shipping addresses to the server as a single request.
 	useEffect( () => {
-		const shouldUpdateBilling = shouldUpdateAddressStore(
-			previousCustomerData.current.billingData,
-			debouncedCustomerData.billingData
-		);
-		const shouldUpdateShipping = shouldUpdateAddressStore(
-			previousCustomerData.current.shippingAddress,
-			debouncedCustomerData.shippingAddress
-		);
-		if ( shouldUpdateBilling || shouldUpdateShipping ) {
-			// Keep refs in sync to avoid multiple calls to API if debounce triggers again.
-			previousCustomerData.current = debouncedCustomerData;
-
-			const requestData = {};
-
-			if ( shouldUpdateBilling ) {
-				requestData.billing_address = {
-					...debouncedCustomerData.billingData,
-					phone: debouncedCustomerData.billingData.phone || null,
-					email: debouncedCustomerData.billingData.email || null,
-				};
-			}
+		if ( ! isShallowEqual( currentBillingData.current, cartBillingData ) ) {
+			currentBillingData.current = cartBillingData;
+		}
+	}, [ cartBillingData ] );
 
-			if ( shouldUpdateShipping ) {
-				requestData.shipping_address =
-					debouncedCustomerData.shippingAddress;
-			}
+	useEffect( () => {
+		if (
+			! isShallowEqual(
+				currentShippingAddress.current,
+				cartShippingAddress
+			)
+		) {
+			currentShippingAddress.current = cartShippingAddress;
+		}
+	}, [ cartShippingAddress ] );
 
-			removeNotice( 'address' );
-			updateCustomerData( requestData ).catch( ( error ) => {
-				addErrorNotice( error.message, {
-					id: 'address',
-				} );
-			} );
+	useEffect( () => {
+		if (
+			! (
+				shouldUpdateAddressStore(
+					currentBillingData.current,
+					debouncedCustomerData.billingData
+				) ||
+				shouldUpdateAddressStore(
+					currentShippingAddress.current,
+					debouncedCustomerData.shippingAddress
+				)
+			)
```

Comment:
Is this still needed if the `debouncedCustomerData` is already doing the equality check? Could you just rely on only updating customer data if `debouncedCustomerData` changes (which would be a dependency change)?

Reply:
Yes this is needed. 

`debouncedCustomerData ` stores all customer data debounced, using shallow equals. 

`shouldUpdateAddressStore` on the other-hand only looks at critical address fields (country, state, postcode) which need pushing to the server. 

Removing this would result in more unnecessary server requests.